### ISSUE TITLE
Redundant OptionalParameterFilter::filter calls

### ISSUE THREAD

**Issue by pine3ree** [laminas/laminas-hydrator#51]
In line:
https://github.com/laminas/laminas-hydrator/blob/cd865fcc01fb5eef62018294456161dc4eb22538/src/ClassMethodsHydrator.php#L166

`$filter` may be either the hydrator's composite filter (which (AND)includes an OptionalParameterFilter) or an ad-hoc composite filter for the current FilterProviderInterface object instance, as provided by the `initCompoisteFilter` call.

This latter ad-hoc filter does not (AND)compose an optional-parameter filter, so the $this->callableMethodFilter->filter call in line: 
https://github.com/laminas/laminas-hydrator/blob/cd865fcc01fb5eef62018294456161dc4eb22538/src/ClassMethodsHydrator.php#L167
is necessary.

At the same time, this is a duplicate check for the internal composite filter case, which already composes it (Just `var_dump($property)` in `OptionalParameterFilter::filter()` to see the consecutive duplicate calls).

Is there a reason for it, or am I just missing some very obvious point? :-)
Couldn't we just add an optional-parameter-filter (even the same shard instance) in the AND filters array of the FilterProviderInterface object in line:
https://github.com/laminas/laminas-hydrator/blob/cd865fcc01fb5eef62018294456161dc4eb22538/src/ClassMethodsHydrator.php#L196
?
like...
```php
    private function initCompositeFilter(object $object) : Filter\FilterComposite
    {
        if ($object instanceof Filter\FilterProviderInterface) {
            return new Filter\FilterComposite(
                [$object->getFilter()],
                [
                     new Filter\MethodMatchFilter('getFilter'),
                     new Filter\OptionalParametersFilter() // ...and dropping the extra call. 
                     // We could even use 1 shared instance of this filter for the insternal composite-filter and all the FilterProviderInterface objects
                ]
            );
        }

        return $this->getCompositeFilter();
    }
```

kind regards

**Comment by pine3ree** [comment]
...or simply just drop
```php
        $compositeFilter->addFilter(
            'parameter',
            new Filter\OptionalParametersFilter(),
            Filter\FilterComposite::CONDITION_AND
        );
```

keeping the extra call. In this case the unatached filter is already shared.

**Comment by froschdesign** [comment]
The fix was released with [version 4.2.0](https://github.com/laminas/laminas-hydrator/releases/tag/4.2.0).
